lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+    #
+    var upperPart: uint64 = 0
+    var lowerPart: uint64 = 0
+    var power: uint64 = hundredQuadrillion
+    for index in 0 ..< 17:
+      power = power div 10
+      if digits[index] != 0:  # to speed things up, we avoid multiplication by zero
+        upperPart += digits[index].uint64 * power
+      if digits[index + 17] != 0:
+        lowerPart += (digits[index + 17].uint64 * power) shr 0
+    if upperPart == 0.uint64:
+      #
+      # if the answer fits into the lower 17 digits, we are done already
+      #
+      result.a = 0'u32
+      result.b = 0'u32
+      result.c = (lowerPart shr 32).uint32      
+      result.d = (lowerPart and allOnes).uint32
+    else:
+      #
+      # move into 128 bits (which, at 34 decimals, will fit into 113 bits)
+      #
+      # four digit equivalent of algorithm using bytes and 100 as splitter:
+      #    digits = [9, 4, 8, 7]
+      #    (upperPart, lowerPart) = (94, 87)
+      #    (upperBig, lowerBig) = 94 * 100 = 9400 = 0x03AE = (03, AE) = (3, 174)
+      #    lowerFinal = (87 + 174) mod 256 = 5 # unsigned ints do modulo automatically
+      #    upperFinal = upperBig
+      #    if 5 < 174 or 5 < 87:
+      #      upperFinal += 1  # aka "carry the one"
+      let (upperBig, lowerBig) = multU64(upperPart, hundredQuadrillion)
+      var lowerFinal = lowerPart + lowerBig
+
+      var upperFinal = upperBig
+      if (lowerFinal < lowerBig) or (lowerFinal < lowerPart):
+        upperFinal += 1.uint64  # "carry the one" (in binary)
+
+      result.a = (upperFinal shr 32).uint32
+      result.b = (upperFinal and allOnes).uint32
+      result.c = (lowerFinal shr 32).uint32
+      result.d = (lowerFinal and allOnes).uint32
+    #
+    # set combo bits
+    #
+    if negativeFlag:
+      result.a = result.a or signMask
+    # unless invoking the medium/long combo (11), the exponent will overlay 00, 01, or 10 bits 1, 2
+    # in other works, an in-range exponent will never have 11 in those two bits; which is why 11 is a flag
+    let exponentMask = (exponent + bias.int).uint32 shl comboShortExponentShiftR
+    result.a = result.a or exponentMask
+
+proc parseFromString(str: string): (DecimalKind, bool, TempSignificandArray, int) =
+  # used internally to parse a decimal string into a temporarary tuple value.
+  type
+    ParseState = enum
+      psPre,            # we haven't found the number yet
+      psLeadingZeroes,  # we are ignoring any leading zero(s)
+      psLeadingMinus,   # we found a minus sign
+      psIntCoeff,       # we are reading the the integer part of a decimal number (NN.nnn)
+      psDecimalPoint,   # we found a single decimal point
+      psFracCoeff,      # we are reading the decimals of a decimal number  (nn.NNN)
+      psSignForExp,     # we are reading the +/- of an exponent
+      psExp,            # we are reading the decimals of an exponent
+      psDone            # ignore everything else
+
+  const
+    IGNORED_CHARS: seq[char] = @[
+      '_', 
+      ','
+    ]
```

Comment:
```suggestion
  const
    IGNORED_CHARS = ['_', ',']
```
An array should suffice here.

Reply:
Also, a bit of a nit, but per NEP-1 constants shouldn't be written in SHOUTING_CASE

Reply:
fixed in next commit.